### ISSUE TITLE
Performance of image (selfie) segmentation w/- Python and GPU

### ISSUE THREAD

**Issue by stephematician** [google/mediapipe#4650]
### Have I written custom code (as opposed to using a stock example script provided in MediaPipe)

Yes

### OS Platform and Distribution

Ubuntu 22.04

### MediaPipe Tasks SDK version

development 

### Task name (e.g. Image classification, Gesture recognition etc.)

Image segmentation

### Programming Language and version (e.g. C++, Python, Java)

Python 3.10

### Describe the actual behavior

The performance of CPU-based image segmentation (`BaseOptions.Delegate.CPU`) is faster than GPU-based (`BaseOptions.Delegate.GPU`). The CPU-based algorithm used about the same amount of CPU resources as the GPU-based approach.

### Describe the expected behaviour

I expected CPU usage to be lower with the GPU-based approach, and for the GPU-based approach to perform faster.

### Standalone code/steps you may have used to try to get what you need

I have been working on a small Python module to do a "fake" green screen on Linux. The code is available here: <https://gitlab.com/stephematician/linuxgreenscreen>.

To enable GPU I had to build with `MEDIAPIPE_DISABLE_GPU=0` (see also <https://github.com/google/mediapipe/issues/4643> for the other changes for Ubuntu 22.04 build).

The v4l2loopback kernel module is also required. To start a virtual camera device:

```
sudo modprode v4l2loopback devices=1 video_nr=10
```

To run the CPU-based segmentation code (for one minute) in Python:

```py
import time

# for redirecting from webcam to virtual device
from linuxgreenscreen.camera_source import CameraSource
from linuxgreenscreen.green_compositor import GreenCompositor
from linuxgreenscreen.virtual_sink import VirtualSink
from v4l2py.device import PixelFormat

source_width, source_height = (1280, 720)
source_fps        = 30
source_color_code = int(PixelFormat.MJPEG).to_bytes(4, 'little').decode()

source_camera = CameraSource(0, width=source_width, height=source_height,
                             fps=source_fps, in_color_code=source_color_code)
greenify = GreenCompositor(width=source_width, height=source_height,
                           # NOTE: change this to True for GPU-based
                           delegate_to_gpu=False)
sink_camera = VirtualSink(10, width=source_width, height=source_height, fps=1,
                          in_color_code='RGB')

source_camera.start()
sink_camera.start()

grabbed = True
n_composited = 0
ns_composite = 0
ts_start     = time.clock_gettime(time.CLOCK_REALTIME)
while grabbed and time.clock_gettime(time.CLOCK_REALTIME) - ts_start < 60:
    grabbed, frame = source_camera()
    if grabbed:
        ns_decoded = time.clock_gettime_ns(time.CLOCK_REALTIME)
        composited_frame = greenify(frame)
        ns_composite += (
            time.clock_gettime_ns(time.CLOCK_REALTIME) - ns_decoded
        )
        n_composited += 1
        sink_camera.write_buffer(composited_frame)
        # over-rides the FPS control of the sink
        sink_camera.notify()

print('Decoding ms: {}'.format(source_camera.decode_ms()))
print('Compositing ms: {}'.format(ns_composite / (1E6 * n_composited)))

sink_camera.stop()
source_camera.stop()
```

The only line that needs to be changed to run GPU-based is in the construction of `greenify` (see NOTE).

To remove the virtual device:

```sh
sudo rmmod v4l2loopback
```

### Other info / Complete Logs

In terms of the execution speed, for the CPU-based code the output of my little example shows that for each frame the compositing time is:

```
Compositing ms: 23.04404925085519
```

Whereas for GPU-based I get:

```
Compositing ms: 34.2793473386545
```

It does appear that the GPU is being used; for example when the image segmenter is constructed I get output:

```
WARNING: Logging before InitGoogleLogging() is written to STDERR
I20230728 17:53:20.849638 242687 gl_context_egl.cc:84] Successfully initialized EGL. Major : 1 Minor: 5
I20230728 17:53:20.907670 242754 gl_context.cc:342] GL version: 3.2 (OpenGL ES 3.2 NVIDIA 525.125.06), renderer: NVIDIA GeForce RTX 2060 with Max-Q Design/PCIe/SSE2
INFO: Created TensorFlow Lite delegate for GPU.
Found Custom Op: Convolution2DTransposeBias
Found Custom Op: Convolution2DTransposeBias
```

And the output from `nvidia-smi` shows that the GPU is doing some work:

```
Fri Jul 28 17:53:54 2023       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 525.125.06   Driver Version: 525.125.06   CUDA Version: 12.0     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  NVIDIA GeForce ...  Off  | 00000000:01:00.0 Off |                  N/A |
| N/A   48C    P8     6W /  30W |    535MiB /  6144MiB |     20%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|    0   N/A  N/A    101543      G   /usr/lib/xorg/Xorg                  4MiB |
|    0   N/A  N/A    242687    C+G   python3                           528MiB |
+-----------------------------------------------------------------------------+
```

**Comment by stephematician** [comment]
Some extra details that might be useful. I am running on a AMD Ryzen w/- Renoir chipset, and an NVIDIA RTX 2060. To get the GPU to work, I invoke python with:

```sh
__NV_PRIME_RENDER_OFFLOAD=1 python3
```

**Comment by stephematician** [comment]
According to @riverzhou - it is possible to attain a speed up via GPU: <https://github.com/google/mediapipe/issues/1651#issuecomment-1655326919> ... however this is via a build w/- CUDA and not the GPU support via OpenGL ES.

**Comment by kuaashish** [comment]
Hello @yichunk,

Could you please look into this issue? Thank you

**Comment by stephematician** [comment]
Would it help if I provided a simpler example?

**Comment by stephematician** [comment]
I tried to get some detail on where the CPU vs GPU code spends it time via google perf tools. As I'm a bit out of my depth, I can't say I saw anything that was obviously slowing down the GPU versus CPU. Attached the code and the output in case it provides any clues for anyone else.

[pprof_nvgpu.pdf](https://github.com/google/mediapipe/files/12418333/pprof_nvgpu.pdf)
[pprof_cpu.pdf](https://github.com/google/mediapipe/files/12418334/pprof_cpu.pdf)
[pprof_amdgpu.pdf](https://github.com/google/mediapipe/files/12418335/pprof_amdgpu.pdf)
[test_image_and_profile_script.tar.gz](https://github.com/google/mediapipe/files/12418348/test_image_and_profile_script.tar.gz)

```sh
LD_PRELOAD=/lib/x86_64-linux-gnu/libprofiler.so.0 \
    CPUPROFILE=sample.profile.cpu \
    CPUPROFILE_FREQUENCY=1000 \
    /usr/bin/python3 -c 'delegate_to_gpu = False; exec(open("run.py").read())'

LD_PRELOAD=/lib/x86_64-linux-gnu/libprofiler.so.0 \
    CPUPROFILE=sample.profile.amdgpu \
    CPUPROFILE_FREQUENCY=1000 \
    /usr/bin/python3 -c 'delegate_to_gpu = True; exec(open("run.py").read())'

__NV_PRIME_RENDER_OFFLOAD=1 \
    LD_PRELOAD=/lib/x86_64-linux-gnu/libprofiler.so.0 \
    CPUPROFILE=sample.profile.nvgpu \
    CPUPROFILE_FREQUENCY=1000 \
    /usr/bin/python3 -c 'delegate_to_gpu = True; exec(open("run.py").read())'

google-pprof --pdf /usr/bin/python3 sample.profile.cpu > pprof_cpu.pdf
google-pprof --pdf /usr/bin/python3 sample.profile.amdgpu > pprof_amdgpu.pdf
google-pprof --pdf /usr/bin/python3 sample.profile.nvgpu > pprof_nvgpu.pdf
```

**Comment by stephematician** [comment]
I switched to the newer `pprof` and got some more clues as to why the CPU usage might be higher with NVIDIA card and why the performance isn't ideal.

1.  For some reason the NVIDIA driver (I think) hammers clock_gettime - this problem has appeared in other contexts <https://forums.developer.nvidia.com/t/nvidia-x11-driver-busy-polls-kernel-on-clock-gettime-in-a-tight-loop/216722>. There might not be much that can be done except wait for better drivers from NVIDIA.
    
    ![image](https://github.com/google/mediapipe/assets/12695305/5d051f3f-6d9f-46b1-863f-f9dbc6ae011e)
    
2.  I also noticed that a rather decent amount of profile time, 11% of total, is spent creating (and destroying) texture buffers when using NVIDIA GPU (specifically via `Image::ConvertToGpu`). Is this the expected behaviour? The results are even worse for the on board AMD GPU with ~37% of the time spent in `ConvertToGpu`).
    
    ![image](https://github.com/google/mediapipe/assets/12695305/1ed3f563-cfba-4cc6-9afb-26f6f2ed5791)

**Comment by stephematician** [comment]
I experimented with using opencv to resize the input tensor prior to segmentation, and the improvements were substantial. The general gist is something like: 

```py
    # input = ... get frame to segment
    # ... resize prior to calling segment_aync
    cv2.resize(src=input, dsize=(256,144), dst=input_tensor,
                 interpolation=cv2.INTER_NEAREST_EXACT)
    segmenter.segment_async(
        image=mp.Image(mp.ImageFormat.SRGB, data=input_tensor)
    )
```

And then in the callback something like:

```py
    # mask_buffer = np.zeros(input.shape[:2], dtype=np.float32)
    def segment_callback(
        result: mp_ImageSegmenterResult, image: mp_Image,
        timestamp_ms: int
    ):
        cv2.resize(src=result.confidence_masks[0].numpy_view(),
                   dsize=input.shape[1::-1], dst=mask_buffer,
                   interpolation=cv2.INTER_LINEAR)
```

I used the same test image as in https://github.com/google/mediapipe/issues/4650#issuecomment-1689752075 and modifed the code to use opencv resize similar to the above, these are the times taken to segment _including_ the resizing operations.

| resize  | CPU     | AMDGPU  | NVGPU   |
|---------|---------|---------|---------|
|  mp     |  9.35ms | 18.23ms |  8.94ms |
|  opencv |  7.95ms | 11.24ms |  3.67ms |

Using `cv2.INTER_AREA` might provide marginally better segmentation when the input is being resized to 256x144; but it comes at a cost in performance (typically 2-3ms for all opencv measures).

As I suspected... MediaPipe doesn't recycle resources like the input tensor - at least on my system - and a fair bit of time is spent allocating (and deallocating) the resources on the GPU. Reducing that footprint _really_ helps in some cases. I wonder if the downsampling complexity is a contributing factor, too.

**Comment by stephematician** [comment]
It's possible that the resource recycling has been addressed in 698b154 although I haven't tested it.
